addons/audio/alsa.c

```diff
@@ -875,6 +878,94 @@ static void alsa_deallocate_recorder(ALLEGRO_AUDIO_RECORDER *r)
    snd_pcm_close(data->capture_handle);
 }
 
+static void _device_list_dtor(void* value, void* userdata)
+{
+   (void)userdata;
+
+   ALLEGRO_AUDIO_DEVICE* device = (ALLEGRO_AUDIO_DEVICE*)value;
+   al_free(device->name);
+   al_free(device->identifier);
+}
+
+static _AL_LIST* alsa_get_devices(void)
+{
+   if (!device_list) {
+      device_list = _al_list_create();
+
+      int rcard = -1;
+      while(snd_card_next(&rcard) == 0) {
+         if (rcard < 0) return device_list;
+
+         snd_ctl_t *handle;
+         char str[64];
+
+         sprintf(str, "hw:%i", rcard);
+         if (snd_ctl_open(&handle, str, 0) < 0) {
+            return device_list;
+         }
+
+         snd_ctl_card_info_t *card_info;
+         snd_ctl_card_info_alloca(&card_info);
+         if (snd_ctl_card_info(handle, card_info) < 0) {
+            continue;
+         }
+
+         snd_ctl_card_info_get_name(card_info);
+         snd_ctl_card_info_get_id(card_info);
+
+         int dev_num = -1;
+
+         while(1) {
+            if (snd_ctl_pcm_next_device(handle, &dev_num) < 0) {
+               break;
+            }
+
+            if (dev_num < 0) break;
+
+            void **hints;
+            int success = snd_device_name_hint(dev_num, "pcm", &hints);
+            if (success < 0) continue;
+            char**n = (char**)hints;
+               
+            char* identifier = 0;
+            char* name = 0;
+            while(*n != NULL)
+            {
+               identifier = snd_device_name_get_hint(*n, "NAME");
+               name = snd_device_name_get_hint(*n, "DESC");
+                  
+               char* ioid = snd_device_name_get_hint(*n, "IOID");
+               if (ioid == NULL || strcmp(ioid, "Output") == 0) break;
+               n++;
+            }
+
+            if (!identifier || !name) continue;
```

Comment:
Log here too.

Reply:
Also: if this continue branch is taken, will we leak the memory allocated by `snd_device_name_hint` (line 926), since it will skip the `snd_device_name_free_hint` (line 957)